File: python/binary-search/search-in-rotated-sorted-array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left, right = 0, len(nums) - 1
        while left <= right:
            middle = (left + right) // 2
            if nums[middle] == target:
                return middle

            if nums[left] <= target and (nums[middle] > target or nums[middle] < nums[left]):
                right = middle - 1
            elif nums[left] >= target and (nums[middle] > target) and nums[middle] < nums[left]:
                right = middle - 1
            else:
                left = middle + 1

        return -1

File: c++/binary-search/search-in-rotated-sorted-array.cc
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int left = 0;
        int right = nums.size() - 1;

        while (left <= right) {
            int middle = (left + right) / 2;
            
            if (nums[middle] == target) {
                return middle;
            }

            if (nums[middle] < target) {
                if (target > nums[right] && nums[middle] < nums[right]) {
                    right = middle - 1;
                } else {
                    left = middle + 1;
                }
            } else {
                if (nums[right] >= target && nums[middle] > nums[right]) {
                    left = middle + 1;
                } else {
                    right = middle - 1;
                }
            }
        }
        return -1;
    }
};
